pageserver/src/tenant.rs

```diff
@@ -2824,9 +2832,8 @@ mod tests {
 
     #[tokio::test]
     async fn no_duplicate_timelines() -> anyhow::Result<()> {
-        let tenant = TenantHarness::create("no_duplicate_timelines")?
-            .load()
-            .await;
+        let harness = TenantHarness::create("no_duplicate_timelines")?;
```

Comment:
Picking out this test (could be any of them) to demonstrate a problem with this patch:

The repository objects must not outlive the TenantHarness.
Otherwise, the harness's repository directory is used after TenantHarness::drop deletes it.
Critically, this includes the background loops of the tenant.

Before this patch, it wasn't a big problem because the temp directory name wouldn't be reused for another test.

But after this patch, tempfile::TempDir might reuse the temp directory of a previous test.

Race condition:

```
T1: start running
T1: create harness with tempdir /tmp/A
T1: start tenant background loop
T1: run test
T1: drop harness at end of test scope, deletes /tmp/A
T2: start running
T2: observes /tmp/A does not exist
T2: create harness with tempdir /tmp/A
T1: background loop does something in repository dir
T1: tokio::test stops polling runtime on which background loop runs
T1: returns
T2: observes changes that T1 made to /tmp/A, fails spuriously
```

Reply:
I don't understand this race, in particular, how 

> But after this patch, tempfile::TempDir might reuse the temp directory of a previous test.

is possible?
We do 

```
let temp_repo_dir = tempfile::tempdir()?;
let repo_dir = temp_repo_dir.path().join(test_name);
```

which should give us different paths, given that the test names are different.

Reply:
Fair, I didn't consider that we append the test name to it.

It's the developer's responsibility to ensure that these are unique, though. Easy mistake when copy-pasting a test. Can't we get rid of that as well?

Idea: Include a UUIDv4 in the tempdir name, a la `tempfile::TempDir::new(format!("pageserver.tenant_harness.repo.{}", gen_uuid_v4()))`.

That way, 
1. devs don't need to keep `test_name` unique. We can get rid of it.
2. with the UUIDv4 in the prefix, we don't need to rely on `tempdir` to not re-use directory names

Reply:
```rust
fn main() {
    let temp_repo_dir = tempfile::tempdir().unwrap();
    dbg!(temp_repo_dir.into_path());
}
```

boils down to https://github.com/Stebalien/tempfile/blob/1c6149d7ce33b2ad8b20a38a8ddce0fc29f68f0e/src/util.rs#L18

and prints on mac:
```
[src/main.rs:3] temp_repo_dir.into_path() = "/var/folders/80/7bn887rd1sj2gt1z64tmrbx80000gn/T/.tmpdIDO7Q"
```
on linux:
```
[src/main.rs:3] temp_repo_dir.into_path() = "/tmp/.tmpm0OMA4"
```

So in terms of races and uniqueness, we should be fine.
Even `TempDir` drops are fine too, since they delete the randomly generated, exclusive, subdirs only.

Specifying the name in general seems redundant indeed, I've removed that.
As a bonus, I've also enabled logs in such tests.

Reply:
> So in terms of races and uniqueness, we should be fine.
> Even TempDir drops are fine too, since they delete the randomly generated, exclusive, subdirs only.

Yeah, and after that, the name can be re-used.
At least on Linux, that's definitely possible.

So, if the path is remembered in some leaked `async task`, we could theoretically hit the race condition outlined earlier.

Reply:
How probable is this to you?
To me, `.tmpm0OMA4` looks like a good enthropy for the amount of tests we have.
I also think, tempfile is the place to solve this, not here.

Reply:
Allright.
Docs say it's 6 bytes of entropy for the name. That's less than a 128bit UUID but probably enough :) 

https://docs.rs/tempfile/latest/tempfile/struct.Builder.html#method.rand_bytes